src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes implementation for the NAT64 translator.
+ *
+ */
+
+#include "nat64_translator.hpp"
+
+#if OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE
+
+#include <openthread/border_router.h>
+#include <openthread/logging.h>
+
+#include "common/code_utils.hpp"
+#include "common/locator_getters.hpp"
+#include "common/uptime.hpp"
+#include "net/checksum.hpp"
+#include "net/ip4_types.hpp"
+#include "net/ip6.hpp"
+
+namespace ot {
+namespace BorderRouter {
```

Comment:
We can do so, however, the NAT64 is only meanful on Border Routers.

- If we allow the user to specify a NAT64 prefix, then the border routing might be work incorrectly since the advertised prefix might be different than the one in effect.
  - If the user sets a prefix, and BORDER_ROUTING is enabled, that means we cannot be a NAT64 capable BR unless we are configured as "auto prefix discovery"
- The next hop (or previous hop) of the message will never in thread network (since thread is IPv6 only) the NAT64 will only work on a border router.
  - That means the router already has IPv4 access, and it does not need a NAT64 translator.

Seems the only possible case is to be a standalone NAT64 translator, but I don't think it is our priority now.

I think it is possible to decouple the NAT64 translator from BR, but I don't think we need to allow users to configure the NAT64 prefix before further dfscussions around these APIs.

Reply:
Yes. I think it is useful to keep different features separate when we can.

The use-case can be that a BR platform/product does not want to use the OT implementation of `RoutingManager` and has its own module for managing the prefixes but may still want to use the OT NAT64 translator feature.

Reply:
Sounds reasonable.

I think we can use a compile time flag in this case? `otNat64SetIp6Prefix` is only available when RoutingManager is not enabled (i.e. BORDER_ROUTERING_ENABLED is not set)

I think this makes sense but providing an API when another API is disabled is somewhat strange, maybe another flag makes more sense?

If we decouple the Nat64Translator from the RoutingManager, I think we can move the implementation to `src/net/nat64.cpp`.

I want to mark this API as a TODO.

Reply:
SGTM. Thanks.